### ISSUE TITLE
Feature checking is slow, even with memoization enabled

### ISSUE THREAD

**Issue by fsateler** [jnunemaker/flipper#613]
Hi,

While deploying flipper in production we encountered some issues, which we think trace back to flipper using a lot of cpu time, even when memoization is active:

```ruby
require 'bundler/inline'

gemfile do
  source 'https://rubygems.org'
  gem 'benchmark-ips', require: false
  gem 'flipper'
end

require 'benchmark/ips'

Flipper.memoize = true
Flipper.preload_all

class Actor
  def flipper_id
    "asdf"
  end
end

$actor = Actor.new

def local_cache feature
  @local_cache ||= {}
  @local_cache.fetch(feature) { @local_cache[feature] = Flipper.enabled? feature, $actor }
end

Benchmark.ips do |x|
  x.report("flipper") { Flipper.enabled?(:test, $actor) }
  x.report("cached") { local_cache(:test) }
  x.compare!
end

```

This results in:

```ruby
Warming up --------------------------------------
             flipper     7.504k i/100ms
              cached   785.639k i/100ms
Calculating -------------------------------------
             flipper     74.828k (± 3.4%) i/s -    375.200k in   5.020694s
              cached      7.871M (± 1.6%) i/s -     40.068M in   5.092171s

Comparison:
              cached:  7870590.8 i/s
             flipper:    74827.8 i/s - 105.18x  (± 0.00) slower

```

That is, flipper takes about 0.01-0.02 milliseconds in evaluating a  feature. This ends up adding a lot of time if the feature is checked in a hot path.

**Comment by jnunemaker** [comment]
I'd be curious about your use case. 

* What type of request/transaction are you using flipper in? 
* How many checks to flipper are happening in a single request/transaction? 
* What is expected response time for those actions and what % of that response time is flipper taking up compared to the rest of the request/transaction?
* What would you expect the timing to be of a single feature enabled check?

In my experience, at GitHub and elsewhere, performance problems were always network calls. 

That said, I haven't worked on optimizing the feature enabled path (in the past handful of years) outside of reducing network calls through memoization and preloading. That seems to be enough for people. 

Another pass of optimizing just the flipper stuff unrelated to network calls is on my list of todos. But it hasn't come up until now so I haven't prioritized it verse other work. 

Look forward to chatting more about this!

**Comment by fsateler** [comment]
Hi! We are still investigating, I'll follow up as we uncover more information. As for your questions:

> What type of request/transaction are you using flipper in?

We were using it to conditionally enable some new permissions. Thus it becomes part of the hot path for web requests.

> How many checks to flipper are happening in a single request/transaction?

Upwards of 20 times (yes, we need to fix that).

> What is expected response time for those actions and what % of that response time is flipper taking up compared to the rest of the request/transaction?

I believe what we are observing is GVL contention: this means other threads cannot run (we use puma).

> What would you expect the timing to be of a single feature enabled check?

A lot closer to a naive memoization strategy, or else it cannot be used in hot paths.

**Comment by jnunemaker** [comment]
> Upwards of 20 times (yes, we need to fix that).

This isn't a problem at all. At GitHub, we did thousands on a single request/transaction. 

> A lot closer to a naive memoization strategy, or else it cannot be used in hot paths.

Are you using ActiveSupport::Notifications for instrumentation? I've seen that become the majority of time before.

**Comment by jnunemaker** [comment]
Oh and to be clear I'm not saying it isn't being a problem for you. I just mean it shouldn't be a problem so we'll have to dig deeper to figure out what is causing your problems. 

What was the response time of the request prior to flipper and after?

**Comment by fsateler** [comment]
It appears that our problem is not directly related to flipper. So, false positive on that part :sweat_smile: . But I still find it weird that flipper is (at least appearing to be) slow. 


> Are you using ActiveSupport::Notifications for instrumentation? I've seen that become the majority of time before.

Indeed! That does become a large part of the time (about half of it!). Disabling the integration has two performance-related benefits: 1. less work done by the instrumentation, and 2. less work done by other notification listeners.

I have profiled with stackprof:

```ruby
StackProf.run(mode: :wall, raw: true) do
   10_000.times { Flipper.enabled?(:test, actor) }
end
```

And I found out that `Flipper::Types::Actor::wrappable?` uses up about 13% of the time it takes to invoke `enabled?`. This in turn is in large part because `respond_to?` is slow for AR models. 

I tried adding a `flipper_id` method to `Flipper::Types::Actor` (just delegating to `thing`), and this appears to cut the time substantially, as we find the `flipper_id` method faster. Removing the explicit check in `#initialize` also helps (we will still fail a few lines later because we invoke `flipper_id`).

Another opportunity would be to somehow cache `Feature#gate_values`. Initializing `GateValues` takes up about 25% of `enabled?` time (after the previous optimizations).


So, in summary:

1. ActiveSupport instrumentation causes significant slowdown. I will probably need to disable it locally.
2. `respond_to?` is sometimes slow, in particular with AR models. Cutting back on the number of invocations would help.
3. Would it be possible to cache the `GateValues` object?

**Comment by jnunemaker** [comment]
1. Yep. I've known it was slow for a while (since the last  I haven't researched why it is slow but I am curious.
2. That's interesting and seems like something we can look into. I'm not sure off hand if I can remove (don't want to break people's stuff) but seems worthwhile. 
3. I don't believe so. Because features are memoized, memoizing gate values would lead to the cache not being updated. I suppose we could do some other shenanigans to make it work. I'll have to think if over. By default I try not to cache any state from the database other than for particular durations of time such as a web request.

I'm going to close this since the problem wasn't flipper, but I do still want to deep dive into your points and see if I can figure something out. I'd also be very open to any explorations via pull request.
